src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs

```diff
@@ -521,6 +512,107 @@ void RewriteConditionToNop (int index)
 				RewriteToNop (index);
 			}
 
+			public void RewriteToNop (int index, int stackDepth)
+			{
+				if (FoldedInstructions == null) {
+					FoldedInstructions = new Collection<Instruction> (Body.Instructions);
+					mapping = new Dictionary<Instruction, int> ();
+				}
+
+				int start_index;
+				for (start_index = index; start_index >= 0 && stackDepth > 0; --start_index) {
+					stackDepth -= GetStackBehaviourDelta (FoldedInstructions[start_index], out bool undefined);
+					if (undefined)
+						return;
+				}
+
+				if (stackDepth != 0) {
+					Debug.Fail ("Invalid IL?");
+					return;
+				}
+
+				while (start_index != index)
+					RewriteToNop (++start_index);
+			}
+
+			static int GetStackBehaviourDelta (Instruction instruction, out bool unknown)
+			{
+				int delta = 0;
+				unknown = false;
+				switch (instruction.OpCode.StackBehaviourPop) {
+				case StackBehaviour.Pop0:
+					break;
+				case StackBehaviour.Pop1:
+				case StackBehaviour.Popref:
+				case StackBehaviour.Popi:
+					--delta;
+					break;
+				case StackBehaviour.Pop1_pop1:
+				case StackBehaviour.Popi_pop1:
+				case StackBehaviour.Popi_popi:
+				case StackBehaviour.Popi_popi8:
+				case StackBehaviour.Popi_popr4:
+				case StackBehaviour.Popi_popr8:
+				case StackBehaviour.Popref_pop1:
+				case StackBehaviour.Popref_popi:
+					delta -= 2;
+					break;
+				case StackBehaviour.Popi_popi_popi:
+				case StackBehaviour.Popref_popi_popi:
+				case StackBehaviour.Popref_popi_popi8:
+				case StackBehaviour.Popref_popi_popr4:
+				case StackBehaviour.Popref_popi_popr8:
+				case StackBehaviour.Popref_popi_popref:
+					delta -= 3;
+					break;
+				case StackBehaviour.Varpop:
+					if (instruction.Operand is IMethodSignature ms) {
+						if (ms.HasThis && instruction.OpCode != OpCodes.Newobj)
+							--delta;
+
+						delta -= ms.Parameters.Count;
+						break;
+					}
+
+					Debug.Fail (instruction.Operand?.ToString ());
+					unknown = true;
+					return 0;
+				default:
+					throw new NotImplementedException (instruction.OpCode.StackBehaviourPop.ToString ());
```

Comment:
I would be hesitant to crash the linker on this - just set unknown to true and return. Maybe Debug.Fail so that we catch this in tests.

Reply:
Main reason - there's no workaround for crashing the linker. If I hit this and the optimization will not work - well not great, but I can still work with my app. If it crashes the linker, I have no way to trim my app.

Reply:
this should happen only for unknown opcodes but addressed anyway.